modules/service/src/main/scala/lucuma/odb/logic/GmosReadoutTime.scala

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2016-2023 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package lucuma.odb.logic
+
+import cats.MonadError
+import cats.Order
+import cats.Order.catsKernelOrderingForOrder
+import cats.syntax.applicative.*
+import cats.syntax.either.*
+import cats.syntax.eq.*
+import cats.syntax.flatMap.*
+import cats.syntax.functor.*
+import cats.syntax.option.*
+import cats.syntax.traverse.*
+import lucuma.core.enums.GmosAmpCount
+import lucuma.core.enums.GmosAmpGain
+import lucuma.core.enums.GmosAmpReadMode
+import lucuma.core.enums.GmosNorthDetector
+import lucuma.core.enums.GmosRoi
+import lucuma.core.enums.GmosSouthDetector
+import lucuma.core.enums.GmosXBinning
+import lucuma.core.enums.GmosYBinning
+import lucuma.core.model.sequence.gmos.DynamicConfig
+import lucuma.core.model.sequence.gmos.GmosCcdMode
+import lucuma.core.util.TimeSpan
+import lucuma.odb.data.Tag
+import lucuma.odb.util.Codecs.*
+import lucuma.odb.util.GmosCodecs.*
+import monocle.Focus
+import monocle.Lens
+import skunk.*
+import skunk.implicits.*
+
+import scala.collection.immutable.SortedMap
+
+
+opaque type GmosReadoutTime = SortedMap[GmosReadoutTime.Key, TimeSpan]
+
+
+object GmosReadoutTime {
+
+  case class Key(
+    detector: Either[GmosNorthDetector, GmosSouthDetector],
+    ccdMode: GmosCcdMode,
+    roi:     GmosRoi
+  )
+
+  object Key {
+    val detector: Lens[Key, Either[GmosNorthDetector, GmosSouthDetector]] =
+      Focus[Key](_.detector)
+
+    val ccdMode: Lens[Key, GmosCcdMode] =
+      Focus[Key](_.ccdMode)
+
+    val roi: Lens[Key, GmosRoi] =
+      Focus[Key](_.roi)
+
+    given Order[Key] =
+      Order.by { a => (
+        a.detector,
+        a.ccdMode,
+        a.roi
+      )}
+
+  }
+
+  extension (m: GmosReadoutTime) {
+
+    def apply(key: GmosReadoutTime.Key): TimeSpan =
+      m.getOrElse(key, TimeSpan.Zero)
+
+    def apply(detector: GmosNorthDetector, gn: DynamicConfig.GmosNorth): TimeSpan =
+      apply(GmosReadoutTime.Key(detector.asLeft, gn.readout, gn.roi))
+
+    def apply(detector: GmosSouthDetector, gs: DynamicConfig.GmosSouth): TimeSpan =
+      apply(GmosReadoutTime.Key(detector.asRight, gs.readout, gs.roi))
+
+  }
+
+  private object statements {
+
+   import lucuma.odb.util.Codecs.*
+   import lucuma.odb.util.GmosCodecs.*
+
+    val select: Query[Void, Tag ~ GmosCcdMode ~ GmosRoi ~ TimeSpan] =
+      sql"""
+        SELECT
+          c_detector,
+          c_x_bin,
+          c_y_bin,
+          c_amp_count,
+          c_amp_gain,
+          c_amp_read_mode,
+          c_roi,
+          c_time
+        FROM
+          t_gmos_readout
+      """.query(tag ~ gmos_ccd_mode ~ gmos_roi ~ time_span)
```

Comment:
Should probably switch to the new Scala 3 representation a la https://github.com/gemini-hlsw/lucuma-odb/pull/431